src/util.tsx

```diff
@@ -1,42 +1,73 @@
-import React, { forwardRef, useMemo, useLayoutEffect, MutableRefObject } from 'react'
-import { Vector2, Object3D } from 'three'
-import { ReactThreeFiber, useThree } from '@react-three/fiber'
-import { Effect, BlendFunction } from 'postprocessing'
-
-type ObjectRef = MutableRefObject<Object3D>
-type DefaultProps = Partial<{ blendFunction: BlendFunction; opacity: number }>
-
-const isRef = (ref: any): ref is ObjectRef => !!ref.current
-
-export const resolveRef = (ref: Object3D | ObjectRef) => (isRef(ref) ? ref.current : ref)
-
-export const wrapEffect = <T extends new (...args: any[]) => Effect>(
-  effectImpl: T,
-  defaultBlendMode: BlendFunction = BlendFunction.NORMAL
-) =>
-  forwardRef<T, ConstructorParameters<typeof effectImpl>[0] & DefaultProps>(function Wrap(
-    { blendFunction, opacity, ...props }: React.PropsWithChildren<DefaultProps & ConstructorParameters<T>[0]>,
-    ref
-  ) {
-    const invalidate = useThree((state) => state.invalidate)
-    const effect: Effect = useMemo(() => new effectImpl(props), [props])
-
-    useLayoutEffect(() => {
-      effect.blendMode.blendFunction = !blendFunction && blendFunction !== 0 ? defaultBlendMode : blendFunction
-      if (opacity !== undefined) effect.blendMode.opacity.value = opacity
-      invalidate()
-    }, [blendFunction, effect.blendMode, opacity])
-    return <primitive ref={ref} object={effect} dispose={null} />
+import * as React from 'react'
+import * as THREE from 'three'
+import { extend, invalidate, ReactThreeFiber } from '@react-three/fiber'
+import type { Effect, BlendFunction } from 'postprocessing'
+
+export const resolveRef = <T extends object>(ref: T | React.MutableRefObject<T>) =>
+  ref && 'current' in ref ? ref.current : ref
+
+export type EffectConstructor = new (...args: any[]) => Effect
```

Comment:
hm might be better using `uknown[]` instead?

Reply:
Apparently can't do that around constructors due to https://github.com/microsoft/TypeScript/issues/37142.